docs/up-to-date-check.md

```diff
@@ -0,0 +1,84 @@
+- :heavy_check_mark: We have received project and item data via Dataflow
+- :heavy_check_mark: The current project's version is up to date with data received via Dataflow
+- :heavy_check_mark: The list of items received via Dataflow has _not_ changed since the last check
+- :heavy_check_mark: `DisableFastUpToDateCheck` is `false` or not specified
+- :heavy_check_mark: No project items have `CopyToOutputDirectory` as `Always`
+
+### Outputs
+
+Output files break down as follows:
+
+`_customOutputs` are `UpToDateCheckOutput` items published via `ProjectSubscription.JointRuleSource`
+`_builtOutputs` are `UpToDateCheckBuilt` items published via `ProjectSubscription.JointRuleSource` (having no or empty `Original` property)
+
+- :heavy_check_mark: All `_customOutputs` and `_builtOutputs` files exist
+
+### Inputs
+
+- :heavy_check_mark: All input files exist
+- :heavy_check_mark: All input files are older than the earliest output (`_customOutputs` and `_builtOutputs`)
+
+### `CopyUpToDateMarker` and `ResolvedCompilationReference`
+
+If `ProjectSubscription.JointRuleSource` published a `CopyUpToDateMarker`...
+
+- :heavy_check_mark: ...at least one `ResolvedCompilationReference` was also published
+- :heavy_check_mark: ...at least one `ResolvedCompilationReference` file exists
+- :heavy_check_mark: ...the `CopyUpToDateMarker` file exists
+- :heavy_check_mark: ...the `CopyUpToDateMarker` file is newer than all existing `ResolvedCompilationReference` files
+
+### Project items with `CopyToOutputDirectory` as `PreserveNewest`
+
+All project items having `CopyToOutputDirectory` as `PreserveNewest`...
+
+- :heavy_check_mark: ...exist
+- :heavy_check_mark: ...have existing output files
+- :heavy_check_mark: ...are older or equal to their output files
+
+### Copied output files
+
+`_copiedOutputFiles` is a map from destination to source paths (relative). It is populated with `UpToDateCheckBuilt` items published via `ProjectSubscription.JointRuleSource` having non-empty `Original` property.
+
+For each `_copiedOutputFiles` source/destination
+
+- :heavy_check_mark: ...the source must exist
+- :heavy_check_mark: ...the destination must exist
+- :heavy_check_mark: ...the destination must be older or the same age as the source
+
+## Implementation
+
+- [`BuildUpToDateCheck.cs`](https://github.com/dotnet/project-system/blob/master/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs)
+- [`BuildUpToDateCheckTests.cs`](https://github.com/dotnet/project-system/blob/master/src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/UpToDate/BuildUpToDateCheckTests.cs)
+
+## Debugging
+
+The `BuildUpToDateCheck` class uses a `BuildUpToDateCheckLogger` to log verbose and info level information that is very useful during debugging.
+
+The log level is obtained via `IProjectSystemOptions.GetFastUpToDateLoggingLevelAsync`.
+
+TODO document how to view this log
```

Comment:
https://github.com/dotnet/project-system/blob/master/docs/repo/debugging/logging.md documents this already.

Reply:
I don't see any logging related to the up to date check in the "Project" output panel. I'll have a look at CPS to see what it does with the `TextWriter` it passes on the interface.

Reply:
Am looking into this and will update once resolved.